src/scene/world-clusters.js

```diff
@@ -0,0 +1,663 @@
+            if (light.enabled && light.type !== LIGHTTYPE_DIRECTIONAL && light.visibleThisFrame) {
+
+                // within light limit
+                if (lightIndex < this.maxLights) {
+
+                    // reuse allocated spot
+                    let clusteredLight;
+                    if (lightIndex < usedLights.length) {
+                        clusteredLight = usedLights[lightIndex];
+                    } else {
+                        // allocate new spot
+                        clusteredLight = new ClusterLight();
+                        usedLights.push(clusteredLight);
+                    }
+
+                    // store light properties
+                    clusteredLight.light = light;
+                    light.getBoundingBox(tempBox);
+                    clusteredLight.min.copy(tempBox.getMin());
+                    clusteredLight.max.copy(tempBox.getMax());
+
+                    lightIndex++;
+                } else {
+                    console.warn("Clustered lighting: more than " + (this.maxLights - 1) + " lights in the frame, ignoring some.");
+                    break;
+                }
+            }
+        }
+
+        usedLights.length = lightIndex;
+    }
+
+    // evaluate the area all lights cover
+    evaluateBounds() {
+
+        const usedLights = this._usedLights;
+
+        // bounds of the area the lights cover
+        const min = this.boundsMin;
+        const max = this.boundsMax;
+
+        // if at least one light (index 0 is null, so ignore that one)
+        if (usedLights.length > 1) {
+
+            // AABB of the first light
+            min.copy(usedLights[1].min);
+            max.copy(usedLights[1].max);
+
+            for (let i = 2; i < usedLights.length; i++) {
+
+                // expand by AABB of this light
+                min.min(usedLights[i].min);
+                max.max(usedLights[i].max);
+            }
+        } else {
+
+            // any small volume if no lights
+            min.set(0, 0, 0);
+            max.set(1, 1, 1);
+        }
+
+        // bounds range
+        this.boundsDelta.sub2(max, min);
+    }
+
+    // evaluate ranges of variables compressed to 8bit texture to allow their scaling to 0..1 range
+    evaluateCompressionLimits(gammaCorrection) {
+
+        let maxAttenuation = 0;
+        let maxColorValue = 0;
+
+        const usedLights = this._usedLights;
+        for (let i = 1; i < usedLights.length; i++) {
+            const light = usedLights[i].light;
+            maxAttenuation = Math.max(light.attenuationEnd, maxAttenuation);
+
+            const color = gammaCorrection ? light._linearFinalColor : light._finalColor;
+            maxColorValue = Math.max(color[0], maxColorValue);
+            maxColorValue = Math.max(color[1], maxColorValue);
+            maxColorValue = Math.max(color[2], maxColorValue);
+        }
+
+        // increase slightly as compression needs value < 1
+        this._maxAttenuation = maxAttenuation + epsilon;
```

Comment:
wouldn't also storing mins be beneficial?

Reply:
it would, but I'm not sure the benefit would be worth extra add instruction per light .. but I'll keep it in mind should precision would need to be improved.